### DISCUSSION TITLE
Best way to connect to home assistant.

### DISCUSSION THREAD

**Discussion by clinta** [pvvx/ATC_MiThermometer#456] [Q&A] [answered]
There is a lot of important information spread across various issues which I've tried to read and understand, but I'm still struggling to figure out the best way to connect these sensors to home assistant and get coverage in a large house.

Here are the options I've seen, and the problems with them:
* ESP32 bluetooth proxies: 
  * https://github.com/pvvx/ATC_MiThermometer/discussions/328
  * Miss measurements
  * Increase battery usage of the sensors, increase power usage of the esp32
  * Does not work with Bluetooth 5.0
* RTL8761 using the Bluetooth Integration in Home Assistant
  * Does not support bluetooth 5.0 
    * https://github.com/pvvx/ATC_MiThermometer/blob/ad46f4d36cbf3417829b39cdbffd95320cba35ae/README.md?plain=1#L72
  * Maybe you can turn on blueooth 5? I could not quite follow these instructions, possibly something lossed in using google translate. It's not clear if this needs to be run every boot before home assistant starts. Not clear which mode I should use.
    * https://github.com/pvvx/ATC_MiThermometer/issues/297#issue-1589485759
  * Trashes your SSD
    * https://github.com/pvvx/ATC_MiThermometer/issues/394
    * Maybe can avoid using passive scanning, but now BTHome does not work.
      * https://github.com/home-assistant/core/issues/76540#issuecomment-1303606132
* RTL8761 using Passive BLE Custom Component
  * Does not support Bluetooth 5.0
    * https://github.com/pvvx/ATC_MiThermometer/blob/ad46f4d36cbf3417829b39cdbffd95320cba35ae/README.md?plain=1#L346
* OpenWrt Bluetooth gateway
  * https://github.com/pvvx/ATC_MiThermometer/issues/352#issuecomment-1583057902
  * Seems to need a custom c program to bridge bluetooth to mqtt
  * An interesting recommendation because I though a lot of the problems are caused by the linux kernel and bluez, does openwrt even solve those issues?
* Use android to bridge bluetooth to mqtt
  * https://github.com/pvvx/ATC_MiThermometer/issues/297#issuecomment-1816813108
  * A possible similar english app in the play store: https://github.com/theengs/app
  * Ideally I'd find a cheap android box with the a recent bluetooth chip I can put in the center of my house for this task. Interested in any recommendations.
  
Can anyone who is using a significant number of these sensors reliably with home assistant share what method you are using to connect them? Are you using Bluetooth 5.0? Is the setup fragile and does it break anytime home assistant updates?

**Comment by lbarjak** [comment]
My main sys:
only zigbee devices (sensors and plugs) > zigbee2mqtt (git clone, nodejs) >
node-red (git clone, nodejs) and home assistant core (git clone, python
venv), on an X240 notebook
Experimental sys:
zigbee devices > iHost + HAOS
xiaomi smart home hub 2 and wifi, bluetooth devices, integrated into the
HAOS

Clint Armstrong ***@***.***> ezt írta (időpont: 2023. dec.
31., V 22:40):

**Reply by clinta** [reply]
Are your bluetooth devices using this firmware? Do they talk to the xaomi hub? Or to HAOS directly? If to HAOS, which bluetooth adapter do you use? And are you trying to use Bluetooth 5.0?

**Comment by lbarjak** [comment]
Bluetooth devices connect to xiaomi hub.

Clint Armstrong ***@***.***> ezt írta (időpont: 2023. dec.
31., V 23:42):

**Reply by clinta** [reply]
Do you use the ATC firmware on your bluetooth sensors? I'm curious what the reason is for customizing the firmware if you're using it with a xiaomi hub.

**Reply by pvvx** [reply]
Xiaomi Gateway 3 -> ZHA
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/5c75c9c4-5a00-41b8-bc01-a1c724f0cb3b)
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/e4cd1872-4562-4932-9db9-fcee2af38132)
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/bcb61000-1fff-44cc-9be6-e59aac9c91e8)

https://github.com/pvvx/ZigbeeTLc
https://github.com/pvvx/BZdevice

**Comment by lbarjak** [comment]
4 thermometers with original firmware work with the xiaomi hub. I changed
one to zigbee out of curiosity, it works with zigbee2mqtt.

Clint Armstrong ***@***.***> ezt írta (időpont: 2024. jan.
1., H 1:54):

**Comment by pvvx** [comment, marked as answer]
**RTL8761BU accepts BLE BT5.2 advertising.**
Specifically, this is the reception of extended BLE advertising in two types of PHY, according to the Bluetooth standard: 1Mbit and 125Kbit. PHY 125 Kbit is "LE Long Range". To switch to reception in "LE Long Range" you must enter the command `hcitool -i hci1 cmd 08 31 03 05 05`.
But "connect" doesn't work in Bluez.
_In "LE Long Range" the communication range is 2 times greater than in Zigbee._

**Comment by pvvx** [comment]
> RTL8761 using the Bluetooth Integration in Home Assistant
> Does not support bluetooth 5.0
> [ATC_MiThermometer/README.md](https://github.com/pvvx/ATC_MiThermometer/blob/ad46f4d36cbf3417829b39cdbffd95320cba35ae/README.md?plain=1#L72)
> Line 72 in [ad46f4d](https://github.com/pvvx/ATC_MiThermometer/commit/ad46f4d36cbf3417829b39cdbffd95320cba35ae)

Home Assistant in "Bluetooth" integration does **not full** support Bluetooth  5.0 and is not implied. 
1. "Connect" doesn't work in Bluez.
2. "Extended Advertisements" in "LE Long Range" works in HA:BTHome if you switch the adapter with the specified command.
3. [Bleak](https://github.com/hbldh/bleak/issues/1225) doesn't work at all with BT5.0

**Reply by clinta** [reply]
That clarifies the situation. Thanks.

**Comment by pvvx** [comment]
**Enable "LE long Range" for HA:BTHome on RTL8761BU:**

(It has been working for more than a year.)

HA:automation
```
alias: BLE to BT5.0 mode
description: Set BT adapter to BT5.0 mode
trigger:
  - platform: homeassistant
    event: start
  - platform: time_pattern
    hours: "8"
action:
  - service: shell_command.set_ble_bt5
    data: {}
mode: single
```
configuration.yaml:
```
shell_command:
  set_ble_bt5: bash /home/pvvx/.homeassistant/phy_coded.sh
```
phy_coded.sh:
```
#!/bin/bash
hcitool -i hci1 cmd 08 31 03 05 05
hciconfig -a hci1 down
hciconfig -a hci1 up
hcitool -i hci1 cmd 08 31 03 05 05
```

Will not work in docker (HA)!

---

MJWSD05MMC in the neighboring house. Distance 200 meters.
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/45a47796-cb36-4178-93cf-1593e66b159b)

**Reply by clinta** [reply]
Thank you. This is very helpful. Going to start down this path.

**Reply by clinta** [reply]
Why do you need to run this every 8 hours? Does it reset itself sometimes?

**Reply by clinta** [reply]
Another question, in the HA bluetooth settings are you using Passive Scanning? And is that preventing the SSD wear problems?

**Reply by clinta** [reply]
I've been trying to replicate this setup. But once I enable "LE Long Range" in Telink Flasher, I cannot detect the sensor in homeassistant, or even on my Pixel 5 using nRF Connect. I believe my pixel 5 should have bluetooth 5 LE, nRF Connect shows "Long range (PHY Coded) supported: YES". So I'm not sure why it can no longer see the device in a scan.

**Reply by clinta** [reply]
And suddenly it started working in home assistant. It took a long time to pick it up, about 20 minutes, but now it's getting updates almost every 10 seconds. And it's getting them from inside the refrigerator which did not work without long range enabled.

Watching it and sometimes it doesn't get updates for about a minute, so I think some advertisements are being missed. But for the most part it is working well. Going to continue testing.

**Reply by pvvx** [reply]
> Why do you need to run this every 8 hours? Does it reset itself sometimes?

Just in case.

https://www.home-assistant.io/integrations/bluetooth/#realtek-rtl8761bu-adapters
> * Realtek RTL8761BU adapters
> * These adapters do not have a reset pin. If they stop responding, there is currently no way for the kernel to reset them automatically. A generic USB reset for these adapters has been introduced in Linux kernel 6.1 and later.

**Reply by pvvx** [reply]
>Another question, in the HA bluetooth settings are you using Passive Scanning? 

Yes

> And is that preventing the SSD wear problems?

Reduces the number of rewrites.

**Reply by pvvx** [reply]
> Watching it and sometimes it doesn't get updates for about a minute, so I think some advertisements are being missed. But for the most part it is working well. Going to continue testing.

`hcitool -i hci1 cmd 08 31 03 05 05` - the adapter with such settings operates in two modes. It alternates reception in Legcy and Coded S8 PHY.  Coded S8 PHY = LE long Randge.

`hcitool -i hci1 cmd 08 31 03 01 01` - only Legacy (1Mb PHY).
`hcitool -i hci1 cmd 08 31 03 04 04` - only Coded S8 PHY (125Kb PHY, LE Long Range).
`hcitool -i hci1 cmd 08 31 03 07 07` - Legacy (1Mb PHY), 2Mb PHY, 125Kb (Coded S8) PHY.

**Reply by pvvx** [reply]
> I believe my pixel 5 should have bluetooth 5 LE, nRF Connect shows "Long range (PHY Coded) supported: YES". So I'm not sure why it can no longer see the device in a scan.

Change settings in nRFConnect.
Legacy + Ext.Advertising.
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/8f4b8386-28ea-4f31-b1a3-69e8b9b05c70)

**Reply by pvvx** [reply]
> I've been trying to replicate this setup. But once I enable "LE Long Range" in Telink Flasher, I cannot detect the sensor in homeassistant, or even on my Pixel 5 using nRF Connect. 

Web bluetooth API (Chrome) cannot scan BT5.0 devices. The reason is there is no support in Linux/Bluez. Say thank you to Linux, because... it hasn't given this to users since 2014.

https://github.com/WebBluetoothCG/web-bluetooth/blob/main/implementation-status.md#scanning-api

But if the device (running LE Long Range) is already scanned, then in Android Chrome you will be able to connect to it.

Linux does not provide the user with any long-distance connectivity capabilities with standard BLE devices and adapters released since 2014. These are decades-old problems due to incorrect implementation of functions in the kernel and reluctance to support new standards in Bluez.
Contact Linus Torvalds :P

**Reply by clinta** [reply]
Is this why I've seen you mention running two adapters? To avoid alternating between modes. One adapter always in legacy, and one always in LE Long Range mode?

**Reply by pvvx** [reply]
Legacy -> CSR8510A10 (BT4.2) Highest number of advertisements received per second.
LE Long Range -> RTL8761BU (BT5.3)  The best reception with the built-in antenna: sensitivity is better than -108..110 dBm. Realtek updates firmware. The firmware is available for download. Loaded into the adapter's RAM at startup. It is possible to create an alternative firmware if you obtain information from Realtek.

**Comment by ralf-e** [comment]
Is there a known standalone solution for a simple BLE 5 LE -MQTT gateway without having to go via HA or similar.  With esp32 it shouldn't work particularly stable.

**Reply by pvvx** [reply]
Doesn't exist because Linux hasn't updated the kernel and Bluetooth API since 2014. They don’t want to give users long-distance communication in BLE.

New bluetooth standards are only supported in Android in the native API.

**Reply by clinta** [reply]
I think a standalone bridge could be made with an android phone running an app like you linked here: https://github.com/pvvx/ATC_MiThermometer/issues/297#issuecomment-1816813108

Or potentially this app for english users: https://play.google.com/store/apps/details?id=com.theengs.app

**Reply by pvvx** [reply]
Can. Supported on Android.
There are many gateways  BLE  for product labels and identifiers, and containers, serving hundreds of such devices over long distances. And there are already ones with [PAwR functions (BT5.4).](https://www.packetcraft.com/bluetooth-pawr)

But (english users) users Linux are limited programmatically only to the ancient standard BT4.2. (With the exception of some Chinese distributions, where patches are applied to the kernel.) :P

**Comment by lbarjak** [comment]
I'm using zigbee2mqtt and node-red, HA just displays, as I wrote before.

Clint Armstrong ***@***.***> ezt írta (időpont: 2024. jan.
1., H 18:07):
